libazureinit/src/provision/ssh.rs

```diff
@@ -67,20 +68,61 @@ pub(crate) fn get_authorized_keys_path_from_sshd() -> io::Result<Option<PathBuf>
     Ok(None)
 }
 
+pub(crate) fn update_sshd_config() -> io::Result<()> {
+    let sshd_config_path = "/etc/ssh/sshd_config";
+    let temp_sshd_config_path = "/etc/ssh/sshd_config.tmp";
+
+    let mut file_content = String::new();
+    {
+        let mut file = OpenOptions::new().read(true).open(sshd_config_path)?;
+        file.read_to_string(&mut file_content)?;
+    }
+
+    let re =
+        Regex::new(r"(?m)^\s*#?\s*PasswordAuthentication\s+no\s*$").unwrap();
```

Comment:
Not sure if a plain `unwrap` is a good idea here in context of the main execution path.
But I see that replacing unwrap with `?` results in another build failures. :thinking:

Reply:
So, the regex call can fail for two reasons. The first is if the pattern is invalid, and secondly if the pattern would produce a regular expression above the configured size limit (and the default is said to be "reasonable").

Given that, as long as we have a test that builds this regex, we can feel okay about unwrapping the error, although I think instead of `unwrap()` we should use `expect("The regular expression is invalid or exceeds the default regex size")`.

The second thing is that regular expression compilation is relatively expensive, so the typical pattern when building a regex from a static string like this is to make it static so it's compiled only once. Now, we're only going to ever call this function once, so in practice it makes no difference from a performance perspective, _but_ if we do that pattern it becomes very easy to write a test specifically for the potential errors we're unwrapping.

The pattern looks like this with the standard library's [LazyLock](https://doc.rust-lang.org/std/sync/struct.LazyLock.html):
```rust
use std::sync::LazyLock;

static PASSWORD_REGEX: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"(?m)^\s*#?\s*PasswordAuthentication\s+no\s*$").unwrap()
}); 
```

Unfortunately, LazyLock is new in Rust 1.80 so we will have to use the `lazy_static` crate instead until our Minimum Supported Rust Version is 1.80+. Doing it with lazy_static is [very similar](https://docs.rs/lazy_static/latest/lazy_static/#syntax).

Reply:
Done